src/app/CommandHandlerInterface.h

```diff
@@ -0,0 +1,158 @@
+#include <lib/core/CHIPTLV.h>
+#include <lib/core/Optional.h>
+
+namespace chip {
+namespace app {
+
+/*
+ * This interface permits applications to register a server-side command handler
+ * at run-time for a given cluster. The handler can either be configured to handle all endpoints
+ * for the given cluster or only handle a specific endpoint.
+ *
+ * If a command is not handled through this interface, it will default to invoking the generated DispatchSingleClusterCommand
+ * instead.
+ *
+ */
+class CommandHandlerInterface
+{
+public:
+    struct HandlerContext
+    {
+    public:
+        HandlerContext(CommandHandler & commandHandler, const ConcreteCommandPath & requestPath, TLV::TLVReader & aReader) :
+            mCommandHandler(commandHandler), mRequestPath(requestPath), mPayload(aReader)
+        {}
+
+        void SetCommandHandled() { mCommandHandled = true; }
+
+        CommandHandler & mCommandHandler;
+        const ConcreteCommandPath & mRequestPath;
+        TLV::TLVReader & mPayload;
+        bool mCommandHandled = false;
+    };
+
+    /**
+     * aEndpointId can be Missing to indicate that this object is meant to be
+     * used with all endpoints.
+     */
+    CommandHandlerInterface(Optional<EndpointId> aEndpointId, ClusterId aClusterId) :
+        mEndpointId(aEndpointId), mClusterId(aClusterId)
+    {}
+
+    virtual ~CommandHandlerInterface() {}
+
+    /**
+     * Callback that must be implemented to handle an invoke request.
+     *
+     * @param [in] handlerContext Context that encapsulates the current invoke request.
+     *                            Handlers are responsible for correctly calling SetCommandHandled()
+     *                            on the context if they did handle the command.
+     *
+     *                            This is not necessary if the HandleCommand() method below is invoked.
+     */
+    virtual void InvokeCommand(HandlerContext & handlerContext) = 0;
+
+    /**
+     * Mechanism for keeping track of a chain of CommandHandlerInterface.
+     */
+    void SetNext(CommandHandlerInterface * aNext) { mNext = aNext; }
+    CommandHandlerInterface * GetNext() const { return mNext; }
+
+    /**
+     * Check whether a this CommandHandlerInterface is relevant for a
+     * particular endpoint+cluster.  An CommandHandlerInterface will be used
+     * for an invoke from a particular cluster only when this function returns
+     * true.
+     */
+    bool Matches(EndpointId aEndpointId, ClusterId aClusterId) const
+    {
+        return (!mEndpointId.HasValue() || mEndpointId.Value() == aEndpointId) && mClusterId == aClusterId;
+    }
+
+    /**
+     * Check whether an CommandHandlerInterface is relevant for a particular
+     * specific endpoint.  This is used to clean up overrides registered for an
+     * endpoint that becomes disabled.
+     */
+    bool MatchesExactly(EndpointId aEndpointId) const { return mEndpointId.HasValue() && mEndpointId.Value() == aEndpointId; }
```

Comment:
Shouldn't a method like this also check mClusterId match?

Reply:
OK, I think that maybe a better suggestion is just to call this MatchesEndpoint, as this explicitly does not take cluster ID into account and I think that is the intent.

```suggestion
    bool MatchesEndpoint(EndpointId aEndpointId) const { return mEndpointId.HasValue() && mEndpointId.Value() == aEndpointId; }
```